File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length];
        int[] prefix = new int[nums.length];
        int[] suffix = new int[nums.length];
        
        prefix[0] = nums[0];
        suffix[nums.length - 1] = nums[nums.length - 1];
        for (int i = 1; i < nums.length; i++) {
            prefix[i] = prefix[i - 1] * nums[i];
            suffix[nums.length - i - 1] = suffix[nums.length - i] * nums[nums.length - i - 1];
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (i == 0) {
                answer[i] = suffix[i + 1];
            }
            else if (i == nums.length - 1) {
                answer[i] = prefix[i - 1];
            }
            else {
               answer[i] = prefix[i - 1] * suffix[i + 1];  
            }
        }
        return answer;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = [nums[0]] + [0] * (n - 1)
        # 1 * 2 * 3 * 4
        # (1 * 2 * 3 * 4) // (1 * 2) = (3 * 4)
        for i in range(1, n):
            prefix[i] = prefix[i - 1] * nums[i]
        answer = [0] * n
        for i in range(n):
            if nums[i] == 0:
                suffix = 1
                for j in range(i + 1, n):
                    suffix *= nums[j]
                answer[i] = prefix[i - 1] * suffix if i > 0 else suffix
            elif prefix[i] == 0:
                answer[i] = 0
            else:
                answer[i] = prefix[i - 1] * (prefix[-1] // prefix[i]) if i > 0 else prefix[-1] // prefix[i]
        return answer